### PR TITLE
[pytorch] Disable fast path in MultiheadAttention in Export

### PR DESCRIPTION
**Pull request by guangy10**
Summary:
We are seeing `aten._native_multi_head_attention` op (not in core Aten op set) is left in the exported graph and causes problems in the downstream at runtime.

Two proposed solutions:
 1. Disable fast path while tracing to leverage the non-optimized path to get decomp, that way, the blamed op won't show up in the exported graph
 2. Add a decomp rule for `aten._native_multi_head_attention`

After discussing with kimishpatel and bdhirsh, #1 is preferred and verified it could immediately unblock the critical model enablement work for PP.

Test Plan: CI

Differential Revision: D48169806

### CODE REVIEW

**Review by kimishpatel — APPROVED** [review]

**Comment by guangy10** [comment]
Fixed failures in test_transformers and test_jit_legacy

**Comment by guangy10** [comment]
Verified that these tests are broken in the base/trunk.
```
FAILED [0.0076s] test/test_autograd.py::TestAutogradFallback::test_base_does_not_require_grad_mode_nothing - RuntimeError: new_fn INTERNAL ASSERT FAILED at "/home/guangyang/pytorch/torch/csrc/autograd/variable.cpp":176, please report a bug to PyTorch.
FAILED [0.0010s] test/test_autograd.py::TestAutogradFallback::test_base_does_not_require_grad_mode_warn - RuntimeError: This is not allowed since there's already a kernel registered from python overriding foo's behavior for CPU dispatch key and _test_autograd_fallback namespace.
FAILED [0.0009s] test/test_autograd.py::TestAutogradFallback::test_composite_registered_to_cpu_mode_nothing - RuntimeError: This is not allowed since there's already a kernel registered from python overriding foo's behavior for CPU dispatch key and _test_autograd_fallback namespace.
FAILED [0.0030s] test/test_autograd.py::TestAutogradFallback::test_composite_registered_to_cpu_mode_warn - RuntimeError: Tried to register an operator (_test_autograd_fallback::foo(Tensor self) -> Tensor) with the same name and overload name multiple times. Each overload's schema should on...
```

**Comment by guangy10** [comment]
Verified those tests are broken on base/trunk.
```
FAILED [0.0586s] test/dynamo/test_dynamic_shapes.py::DynamicShapesReproTests::test_dynamic_shapes_float_guard_dynamic_shapes - Failed: Unexpected success
FAILED [0.6717s] test/dynamo/test_dynamic_shapes.py::DynamicShapesAotAutogradFallbackTests::test_aot_sequence_nr - AssertionError: 'SeqN[542 chars]aten.ones_like.default|\n12|aten.expand.defaul[368 chars]t|\n' != 'SeqN[542 chars]aten.expand.default|\n12|aten.div.Scalar|\n11|[340 chars]t|\n'
FAILED [0.4966s] test/dynamo/test_dynamic_shapes.py::DynamicShapesAotAutogradFallbackTests::test_aot_sequence_nr_dynamic_shapes - AssertionError: 'SeqN[542 chars]aten.ones_like.default|\n12|aten.expand.defaul[368 chars]t|\n' != 'SeqN[542 chars]aten.expand.default|\n12|aten.div.Scalar|\n11|[340 chars]t|\n'
```

Those tests are actually flaky tests. If run it one-by-one, you may see it pass sometimes. Anyway, failures are not relevant to this PR

**Comment by guangy10** [comment]
The distributed/test_distributed_spawn.py failures doesn't seem to be relevant. 

@pytorchbot merge -i

**Comment by mikekgfb** [comment]
the Inference Fastpath context manager in https://github.com/pytorch/pytorch/pull/107014 might offer a more general solution to controlling what kernels are being used (similar to the sdp_kernel context manager)

File: torch/nn/modules/activation.py
Comment (albanD) [inline torch/nn/modules/activation.py:898]: Why is this code in torch.nn ?
That doesn't look like the right place to put it and there are 100% chance this is going to get broken if proxy mode is modified as no-one will expect to update this file.
Could you please move this utility within fx?

Comment (albanD) [reply torch/nn/modules/activation.py:898]: FYI @bdhirsh

Comment (guangy10) [reply torch/nn/modules/activation.py:898]: Yeah, it is a hack suggested by @bdhirsh to temporarily unblock Executorch during model tracing. #107014 seems to plan to handle this in a nicer way.

Comment (albanD) [reply torch/nn/modules/activation.py:898]: That other PR is not removing this utility though?

**Comment by mikekgfb** [comment]
Maybe the right answer for this fine grained control of what algorithms to use is a connect manager similar to what we already do for sdpa?

See #107163 for a possible implementation

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: torch/nn/modules/activation.py
Comment (kimishpatel) [reply torch/nn/modules/activation.py:898]: Agree with @albanD we should mvoe this within fx. @albanD any suggestions on where it can go?

Comment (albanD) [reply torch/nn/modules/activation.py:898]: I will defer to @bdhirsh for the final answer but next to the make_fx API would be my first suggestion

**Pull request closed without merging**
